pkg: Add script to check third party package version

namberino opened the pull request:
### Contribution description

This provides a script that will read the Makefile of third party packages in the `pkg/` directory and compare the `PKG_VERSION` hash with the latest upstream release hash or commit hash and check if the package is outdated or not.

The script requires a GitHub API token to run. 

### Testing procedure

Running the script and inputing a GitHub API token should print out a list of outdated packages.

### Issues/PRs references

Fixes #13911

mguetschow reviewed (changes requested):
Thanks, that seems to be a nice addition!

From a very quick glance at the changes, already some comments:
- could you adapt the script to follow our [python coding convention](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#python-coding-convention)? In particular, provide and use a `main` function, add a copyright notice and shorten lines longer than 120 chars.
- could you move the script under `dist/tools/pkg_version_check` (or similar) and provide a (short) README explaining its function?

mguetschow commented on pkg/version_check.py:
```diff
@@ -0,0 +1,173 @@
+    """
+    file_contents = {}
+    # Get the current working directory
+    current_dir = os.getcwd()
```
I think it would be better to stay independent of the cwd.

mguetschow commented:
Tried it locally, the python file is missing the executable rights (`chmod u+x version_check.py`).

```
Total number of outdated packages: 74
```
:eyes: 

I think it would be nice to provide more concise output, only listing failures and outdated packages.

namberino commented:
I've moved the script to `dist/tools` and reformatted the output so that it will only print out the outdated and failed packages. I've also added a README for this script and rewrote the script to follow RIOT's coding convention

mguetschow commented on dist/tools/pkg_version_check/README.md:
```diff
@@ -0,0 +1,13 @@
+# pkg version check script
+
+This directory contains the script used for checking for outdated packages in the pkg directory.
```
```suggestion
This script checks for outdated packages in the `pkg` directory.
```

mguetschow commented on dist/tools/pkg_version_check/README.md:
```diff
@@ -0,0 +1,13 @@
+
+The script will get the `PKG_VERSION` and `PKG_URL` in a package Makefile. Then it will get the latest release hash of the package on GitHub and compare that with the `PKG_VERSION` hash. If it cannot find any release for the package on GitHub, it will fallback to getting the latest commit hash and compare that with the `PKG_VERSION` hash.
+
+Currently, the script can only check packages that is hosted on GitHub.
```
```suggestion
Currently, the script can only check packages that are hosted on GitHub.
```

mguetschow commented on dist/tools/pkg_version_check/README.md:
```diff
@@ -0,0 +1,13 @@
+
+The script will check every packages inside the `pkg` directory, then output the total number of outdated packages, the name of the outdated packages, the number of packages the script cannot check, and the name of the packages the script cannot check.
+
+Once the script is ran, we'll need to input a GitHub API token. 
```
```suggestion
A GitHub API token is required to retrieve information from Github. The script will ask for the API token upon execution.
```

mguetschow commented on dist/tools/pkg_version_check/version_check.py:
```diff
@@ -124,50 +108,88 @@ def get_release_or_commit_hash(repo_url):
         # Submit tasks to the executor
         for key, item in versions.items():
-            future = executor.submit(check_repository, key, item)
+            future = executor.submit(check_repository, key, item, 
```
```suggestion
            future = executor.submit(check_repository, key, item,
```

mguetschow commented on dist/tools/pkg_version_check/version_check.py:
```diff
@@ -124,50 +108,88 @@ def get_release_or_commit_hash(repo_url):
         for key, item in versions.items():
-            future = executor.submit(check_repository, key, item)
+            future = executor.submit(check_repository, key, item, 
+                                     api_url, base_url, headers, lock, 
```
```suggestion
                                     api_url, base_url, headers, lock,
```

mguetschow commented on dist/tools/pkg_version_check/version_check.py:
```diff
@@ -124,50 +108,88 @@ def get_release_or_commit_hash(repo_url):
+    print("\nPackages that couldn't be checked: ")
+    for package in failures:
+        print(package)
 
```
```suggestion


```

mguetschow commented on dist/tools/pkg_version_check/version_check.py:
```diff
@@ -66,50 +59,41 @@ def get_makefiles():
-    If there is no release, get the latest commit hash from the main branch, and then master branch if necessary.
-    """
+def get_release_or_commit_hash(repo_url, headers):
+    # Get the commit hash
```
You could have left the function documentation in a docstring, I don't think that'd be against the coding conventions.

mguetschow reviewed (changes requested):
Thanks, looks mostly good to me. Below just some suggestions and minor nits.

namberino commented:
Thanks for the suggestions, I've the script and the README

mguetschow commented on dist/tools/pkg_version_check/version_check.py:
```diff
@@ -0,0 +1,196 @@
+                        # Extract the value of PKG_URL
+                        pkg_url = line.strip().split("=")[1]
+                        # Process space and comments
+                        if pkg_url[0] == " ":
+                            pkg_url = pkg_url[1:]
+                        pkg_url = pkg_url.split(" ")[0]
```
```suggestion
                        pkg_url = pkg_url.strip()
```
shouldn't this do the trick already? same below

namberino replied:
I used split because it splits the `PKG_URL` line into an array. This makes it easier to remove the inline comments as I could just select the 0th element of the array to get the `PKG_URL` only.

mguetschow reviewed (approved):
LGTM then! Please squash your commits together into a single one.

namberino commented on dist/tools/bmp/bmp.py:
```diff
@@ -85,7 +85,7 @@ def detect_targets(gdbmi, res):
     while True:
         for msg in res:
             if msg['type'] == 'target':
-                m = re.fullmatch(pattern=r"\s*(\d)+\s*(.*)\\n", string=msg['payload'])
```
Sorry about that, I'll remove the other changes.
